src/v/coproc/router.cc

```diff
@@ -150,46 +189,28 @@ router::process_batch(std::vector<process_batch_request::data> batch) {
       });
 }
 
-ss::future<router::opt_req_data>
-router::route_ntp(const model::ntp& ntp, topic_state& ts) {
-    /**
-     * Making a reader will grab a mutual exclusion lock on reading from the
-     * requested log. This is OK for now since the only topic_ingestion_policy
-     * supported will be 'latest'
-     *
-     * The last offset is recorded and the request is prepared and returned.
-     */
-    return make_reader_cfg(ts.log, ts.head)
-      .then([this, ntp, log = ts.log, sids = ts.scripts](
-              opt_cfg config) mutable {
-          if (!config) {
-              return ss::make_ready_future<opt_req_data>(std::nullopt);
+ss::future<router::opt_req_data> router::route_ntp(
+  const model::ntp& ntp,
+  storage::log_reader_config config,
+  ss::lw_shared_ptr<router_source_manager::topic_state> ts) {
+    return ts->log.make_reader(config)
+      .then([this](model::record_batch_reader reader) {
+          return extract_offset(std::move(reader));
+      })
+      .then([ntp = std::move(ntp),
+             ts](std::optional<offset_rbr_pair> offset_and_batch) mutable {
+          if (!offset_and_batch) {
+              return opt_req_data(std::nullopt);
           }
-          return log.make_reader(*config)
-            .then([this, ntp](model::record_batch_reader reader) {
-                return extract_offset(std::move(reader));
-            })
-            .then([this, ntp, sids = std::move(sids)](
-                    std::optional<offset_rbr_pair> p) {
-                if (!p) {
-                    return opt_req_data(std::nullopt);
-                }
-                auto& [offset, rbr] = *p;
-                auto found = _sources.find(ntp);
-                if (found == _sources.end()) {
-                    vlog(
-                      coproclog.info,
-                      "Ntp removed before batch assemble: {}",
-                      ntp);
-                    return opt_req_data(std::nullopt);
-                }
-                found->second.head.dirty = offset;
-                std::vector<script_id> ids(
-                  std::make_move_iterator(sids.begin()),
-                  std::make_move_iterator(sids.end()));
-                return opt_req_data(process_batch_request::data{
-                  .ids = std::move(ids), .ntp = ntp, .reader = std::move(rbr)});
-            });
+          auto& [offset, rbr] = *offset_and_batch;
+          ts->dirty = offset;
+          std::vector<script_id> ids(
+            std::make_move_iterator(ts->scripts.begin()),
+            std::make_move_iterator(ts->scripts.end()));
+          return opt_req_data(process_batch_request::data{
+            .ids = std::move(ids),
```

Comment:
You could directly construct ids, and given that they're just `int64_t`s, no `move_iterator` required:
```
            .ids{ts->scripts.begin(), ts->scripts.end()},
```
or
```
            .ids = {ts->scripts.begin(), ts->scripts.end()},
```
If you prefer the symmetry.

Reply:
Fire comment , thanks!